### PR TITLE
Ruby family: 2.4.2, 2.3.5, 2.2.8

### PR DESCRIPTION
**Pull request by DomT4**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by ilovezfs** [comment]
Eleven days and a pile more CVEs. :(

File: Formula/ruby@2.3.rb [hunk 1]
```diff
@@ -37,10 +25,14 @@ class RubyAT23 < Formula
...
-  # https://www.ruby-lang.org/en/news/2017/08/29/multiple-vulnerabilities-in-rubygems/
-  patch :p0 do
-    url "https://bugs.ruby-lang.org/attachments/download/6691/rubygems-2613-ruby23.patch"
-    sha256 "510567a43d57ea9c8c7436b14e78d0a4d33380f410443dcf350b9867c9745748"
+  # Reverts an upstream commit which incorrectly tries to install headers
+  # into SDKROOT, if defined
+  # See https://bugs.ruby-lang.org/issues/11881
+  # The issue has been fixed on HEAD as of 1 Jan 2016, but has not been
+  # backported to the 2.3 branch yet and patch is still required.
+  patch do
```
Comment (ilovezfs) [inline Formula/ruby@2.3.rb]: put it back lol

**Comment by ilovezfs** [comment]
Um so … do we know why we're carrying that patch around anyway when https://bugs.ruby-lang.org/issues/11881 was closed?

File: Formula/ruby@2.3.rb (hunk 1 above)
Comment (DomT4) [reply Formula/ruby@2.3.rb]: But ILZ :sob:

**Comment by DomT4** [comment]
> Eleven days and a pile more CVEs. :(

Yup. I love it when major CVEs build up & upstreams are like eh it can probably wait 🤷‍♂️.

> Um so … do we know why we're carrying that patch around anyway when https://bugs.ruby-lang.org/issues/11881 was closed?

That's a good question. Maybe it was applied but the patch kept applying cleanly anyway for some reason. I'll check.

File: Formula/ruby@2.3.rb (hunk 1 above)
Comment (ilovezfs) [reply Formula/ruby@2.3.rb]: We'll need a component order audit for `patch do` before this becomes a reason to move things around. I figure eventually you'll break down and open a PR for that to `brew`.

**Comment by DomT4** [comment]
Hmm. The patch can go I think. https://github.com/ruby/ruby/commit/538ba389836b090be5c913cdfe820d3d3f626e2b is in this release.

**Comment by ilovezfs** [comment]
Lovely

File: Formula/ruby@2.3.rb [hunk 2]
```diff
@@ -117,7 +99,7 @@ def program_suffix
   end
 
   def rubygems_bindir
-    "#{HOMEBREW_PREFIX}/bin"
+    "#{HOMEBREW_PREFIX}/lib/ruby/gems/#{abi_version}/bin"
```
Comment (ilovezfs) [inline Formula/ruby@2.3.rb]: what's going on here?

Comment (DomT4) [reply Formula/ruby@2.3.rb]: It's what we do for all the versioned Rubies, beside this one. I think that was a likely screw up whenever this was added as a versioned formulae.

Comment (ilovezfs) [reply Formula/ruby@2.3.rb]: :( ok

**Review by ilovezfs — APPROVED** [review]

File: Formula/ruby@2.3.rb (hunk 2 above)
Comment (DomT4) [reply Formula/ruby@2.3.rb]: Just prevents the main `ruby` from having its gems accidentally smashed by the older `ruby@whatever` I guess.

**Comment by DomT4** [comment]
You may wish to kill the build sigh. Apparently some things are going to need `revision` bumps because they've attached themselves to the whole `dylib` version:

```
16:22:24 ==> /usr/local/Cellar/vim/8.0.1100/bin/vim -T dumb -s commands.vim test.txt
16:22:24 dyld: Library not loaded: /usr/local/opt/ruby/lib/libruby.2.4.1.dylib
16:22:24   Referenced from: /usr/local/Cellar/vim/8.0.1100/bin/vim
16:22:24   Reason: image not found
16:22:24 Error: vim: failed
```

```
16:22:29 ==> brew install vim@7.4
16:22:32 ==> brew linkage --test vim@7.4
16:22:32 ==> FAILED
16:22:32 Missing libraries:
16:22:32   /usr/local/opt/ruby/lib/libruby.2.4.1.dylib
```

```
16:22:46 `/Users/brew/Jenkins/workspace/Homebrew Core Pull Requests/version/el_capitan/home` is not writable.
16:22:46 Bundler will use `/tmp/bundler/home/brew' as your home directory temporarily.
16:22:46 [ERROR] dlopen(/usr/local/Cellar/wpscan/2.9.3/libexec/gems/ffi-1.9.18/lib/ffi_c.bundle, 9): Library not loaded: /usr/local/opt/ruby/lib/libruby.2.4.1.dylib
16:22:46   Referenced from: /usr/local/Cellar/wpscan/2.9.3/libexec/gems/ffi-1.9.18/lib/ffi_c.bundle
16:22:46   Reason: image not found - /usr/local/Cellar/wpscan/2.9.3/libexec/gems/ffi-1.9.18/lib/ffi_c.bundle
16:22:46 Warning: The post-install step did not complete successfully
16:22:46 You can try again using `brew postinstall wpscan`
```

**Review by ilovezfs — APPROVED** [review]

**Comment by fxcoudert** [comment]
Why does `vim` require revision bump when it depends on `:ruby` and should thus build with system Ruby rather than Homebrew's?

**Comment by DomT4** [comment]
That should be the last of them. Missed a couple because the `linkage` message is a lot less messy than the test failure :smile:

> Why does vim require revision bump when it depends on :ruby and should thus build with system Ruby rather than Homebrew's?

Bottles build against the `default_formula`, always.

**Comment by fxcoudert** [comment]
> Bottles build against the default_formula, always.

I knew that for things that are not provided by the system, but thought if the system provided it (like ruby or perl) then it would not. I fail to see why we would force an install of ruby when the system already has one…

**Comment by DomT4** [comment]
> I knew that for things that are not provided by the system, but thought if the system provided it (like ruby or perl) then it would not.

Nope. There was a bug for a while around this behaviour but the behaviour of building against `default_formula` where there's a requirement used in-formula is [expected behaviour](https://github.com/Homebrew/brew/pull/2969). If you build from source you don't get Homebrew's stuff attached, but if you pour the bottle you do.

**Comment by DomT4** [comment]
If you remember a while back, when `vim` started finally being bottled, people were confused as to why Homebrew's stuff was suddenly attached to a `vim` installation, which is why there's a hack in the `vim` formula to allow people to pass options that ensure `vim` is built from source every time on upgrade without having to pass the `-s` flag.

**Comment by DomT4** [comment]
9 commits later & we are green :tada:.

**Comment by ilovezfs** [comment]
:gem:

**Comment by RandomDSdevel** [comment]
@DomT4, @ilovezfs, @fxcoudert:  I should probably take the time to open a new issue for this, but I just thought I might mention in passing the discovery of https://bugs.ruby-lang.org/issues/13899 in passing (a notice of this was added to [the Ruby 2.4.2 release announcement](https://www.ruby-lang.org/en/news/2017/09/14/ruby-2-4-2-released/) yesterday under the heading 'Known Problem.')

**Comment by DomT4** [comment]
We could apply the simple upstream diff to 2.4.2 & 2.3.5 if ILZ is cool with the idea. Presumably not a ton of people are building `--with-gmp` or we'd get more complaints by now, but, easy fix.

**Comment by ilovezfs** [comment]
Not sure what they're talking about.
```
iMac-TMP:~ joe$ brew linkage ruby
System libraries:
  /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
  /usr/lib/libSystem.B.dylib
  /usr/lib/libffi.dylib
  /usr/lib/libncurses.5.4.dylib
  /usr/lib/libobjc.A.dylib
  /usr/lib/libutil.dylib
  /usr/lib/libz.1.dylib
Homebrew libraries:
  /usr/local/opt/gdbm/lib/libgdbm.4.dylib (gdbm)
  /usr/local/opt/gmp/lib/libgmp.10.dylib (gmp)
  /usr/local/opt/libyaml/lib/libyaml-0.2.dylib (libyaml)
  /usr/local/opt/openssl/lib/libcrypto.1.0.0.dylib (openssl)
  /usr/local/opt/openssl/lib/libssl.1.0.0.dylib (openssl)
  /usr/local/opt/readline/lib/libreadline.7.dylib (readline)
  /usr/local/Cellar/ruby/2.4.2/lib/libruby.2.4.2.dylib (ruby)
Undeclared dependencies with linkage:
  gdbm
```

**Comment by ilovezfs** [comment]
probably our
```
    %w[readline gdbm gmp libffi].each do |dep|
      paths << Formula[dep].opt_prefix if build.with? dep
    end

    args << "--with-opt-dir=#{paths.join(":")}"
```
makes this a non-issue. In any case, I don't see a need to patch here.

**Comment by DomT4** [comment]
Cool. Thanks for checking on that, it was on my list post-go stuff :see_no_evil:.

**Pull request closed without merging**